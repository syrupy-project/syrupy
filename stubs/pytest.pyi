from typing import (
    Any,
    Callable,
    TypeVar,
)

ReturnType = TypeVar("ReturnType")

def fixture(func: Callable[..., ReturnType]) -> Callable[..., ReturnType]: ...

class Function: ...
class Item: ...
