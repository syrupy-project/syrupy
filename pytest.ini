[pytest]
addopts = -p pytester -p syrupy
