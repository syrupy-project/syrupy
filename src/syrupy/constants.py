SNAPSHOT_DIRNAME = "__snapshots__"
SNAPSHOT_EMPTY_FOSSIL_KEY = "empty snapshot fossil"
SNAPSHOT_UNKNOWN_FOSSIL_KEY = "unknown snapshot fossil"

EXIT_STATUS_FAIL_UNUSED = 1

SYMBOL_ELLIPSIS = "..."  # U+2026
SYMBOL_NEW_LINE = "␤"  # U+2424
SYMBOL_CARRIAGE = "␍"  # U+240D

DISABLE_COLOR_ENV_VAR = "ANSI_COLORS_DISABLED"
DISABLE_COLOR_ENV_VARS = {DISABLE_COLOR_ENV_VAR, "NO_COLOR"}

PYTEST_NODE_SEP = "::"
