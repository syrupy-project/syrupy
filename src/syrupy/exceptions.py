class SnapshotDoesNotExist(Exception):
    """Snapshot does not exist"""


class FailedToLoadModuleMember(Exception):
    """Failed to load specific member in a module"""
