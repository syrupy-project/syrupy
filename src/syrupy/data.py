from types import MappingProxyType
from typing import (
    TYPE_CHECKING,
    Dict,
    Iterator,
    List,
    Optional,
)

import attr

from .constants import (
    SNAPSHOT_EMPTY_FOSSIL_KEY,
    SNAPSHOT_UNKNOWN_FOSSIL_KEY,
)

if TYPE_CHECKING:
    from .types import SerializedData


@attr.s(frozen=True)
class Snapshot:
    name: str = attr.ib()
    data: Optional["SerializedData"] = attr.ib(default=None)


@attr.s(frozen=True)
class SnapshotEmpty(Snapshot):
    name: str = attr.ib(default=SNAPSHOT_EMPTY_FOSSIL_KEY, init=False)


@attr.s(frozen=True)
class SnapshotUnknown(Snapshot):
    name: str = attr.ib(default=SNAPSHOT_UNKNOWN_FOSSIL_KEY, init=False)


@attr.s
class SnapshotFossil:
    """A collection of snapshots at a save location"""

    location: str = attr.ib()
    _snapshots: Dict[str, "Snapshot"] = attr.ib(factory=dict)

    @property
    def has_snapshots(self) -> bool:
        return bool(self._snapshots)

    def get(self, snapshot_name: str) -> Optional["Snapshot"]:
        return self._snapshots.get(snapshot_name)

    def add(self, snapshot: "Snapshot") -> None:
        self._snapshots[snapshot.name] = snapshot
        if snapshot.name != SNAPSHOT_EMPTY_FOSSIL_KEY:
            self.remove(SNAPSHOT_EMPTY_FOSSIL_KEY)

    def merge(self, snapshot_fossil: "SnapshotFossil") -> None:
        for snapshot in snapshot_fossil:
            self.add(snapshot)

    def remove(self, snapshot_name: str) -> None:
        self._snapshots.pop(snapshot_name, None)

    def __len__(self) -> int:
        return len(self._snapshots)

    def __iter__(self) -> Iterator["Snapshot"]:
        return iter(self._snapshots.values())


SNAPSHOTS_EMPTY = MappingProxyType({SnapshotEmpty().name: SnapshotEmpty()})
SNAPSHOTS_UNKNOWN = MappingProxyType({SnapshotUnknown().name: SnapshotUnknown()})


@attr.s(frozen=True)
class SnapshotEmptyFossil(SnapshotFossil):
    """This is a saved fossil that is known to be empty and thus can be removed"""

    _snapshots: Dict[str, "Snapshot"] = attr.ib(default=SNAPSHOTS_EMPTY, init=False)

    @property
    def has_snapshots(self) -> bool:
        return False


@attr.s(frozen=True)
class SnapshotUnknownFossil(SnapshotFossil):
    """This is a saved fossil that is unclaimed by any extension currently in use"""

    _snapshots: Dict[str, "Snapshot"] = attr.ib(default=SNAPSHOTS_UNKNOWN, init=False)


@attr.s
class SnapshotFossils:
    _snapshot_fossils: Dict[str, "SnapshotFossil"] = attr.ib(factory=dict)

    def get(self, location: str) -> Optional["SnapshotFossil"]:
        return self._snapshot_fossils.get(location)

    def add(self, snapshot_fossil: "SnapshotFossil") -> None:
        self._snapshot_fossils[snapshot_fossil.location] = snapshot_fossil

    def update(self, snapshot_fossil: "SnapshotFossil") -> None:
        snapshot_fossil_to_update = self.get(snapshot_fossil.location)
        if snapshot_fossil_to_update is None:
            snapshot_fossil_to_update = SnapshotFossil(
                location=snapshot_fossil.location
            )
            self.add(snapshot_fossil_to_update)
        snapshot_fossil_to_update.merge(snapshot_fossil)

    def merge(self, snapshot_fossils: "SnapshotFossils") -> None:
        for snapshot_fossil in snapshot_fossils:
            self.update(snapshot_fossil)

    def __iter__(self) -> Iterator["SnapshotFossil"]:
        return iter(self._snapshot_fossils.values())

    def __contains__(self, key: str) -> bool:
        return key in self._snapshot_fossils


@attr.s
class DiffedLine:
    a: str = attr.ib(default=None)
    b: str = attr.ib(default=None)
    c: List[str] = attr.ib(factory=list)
    diff_a: str = attr.ib(default="")
    diff_b: str = attr.ib(default="")

    @property
    def has_snapshot(self) -> bool:
        return self.a is not None

    @property
    def has_received(self) -> bool:
        return self.b is not None

    @property
    def is_complete(self) -> bool:
        return self.has_snapshot and self.has_received

    @property
    def is_context(self) -> bool:
        return bool(self.c)
