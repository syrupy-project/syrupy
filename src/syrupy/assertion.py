import traceback
from gettext import gettext
from typing import (
    TYPE_CHECKING,
    Any,
    Callable,
    Dict,
    List,
    Optional,
    Type,
)

import attr

from .exceptions import SnapshotDoesNotExist

if TYPE_CHECKING:
    from .extensions.base import AbstractSyrupyExtension
    from .location import PyTestLocation
    from .session import SnapshotSession
    from .types import (
        PropertyFilter,
        PropertyMatcher,
        SerializableData,
        SerializedData,
    )


@attr.s
class AssertionResult:
    snapshot_location: str = attr.ib()
    snapshot_name: str = attr.ib()
    asserted_data: Optional["SerializedData"] = attr.ib()
    recalled_data: Optional["SerializedData"] = attr.ib()
    created: bool = attr.ib()
    updated: bool = attr.ib()
    success: bool = attr.ib()
    exception: Optional[Exception] = attr.ib()

    @property
    def final_data(self) -> Optional["SerializedData"]:
        if self.created or self.updated:
            return self.asserted_data
        return self.recalled_data


@attr.s(cmp=False, repr=False)
class SnapshotAssertion:
    name: str = attr.ib(default="snapshot")
    _session: "SnapshotSession" = attr.ib(kw_only=True)
    _extension_class: Type["AbstractSyrupyExtension"] = attr.ib(kw_only=True)
    _test_location: "PyTestLocation" = attr.ib(kw_only=True)
    _update_snapshots: bool = attr.ib(kw_only=True)
    _exclude: Optional["PropertyFilter"] = attr.ib(
        init=False, default=None, kw_only=True
    )
    _extension: Optional["AbstractSyrupyExtension"] = attr.ib(
        init=False, default=None, kw_only=True
    )
    _executions: int = attr.ib(init=False, default=0, kw_only=True)
    _execution_results: Dict[int, "AssertionResult"] = attr.ib(
        init=False, factory=dict, kw_only=True
    )
    _matcher: Optional["PropertyMatcher"] = attr.ib(
        init=False, default=None, kw_only=True
    )
    _post_assert_actions: List[Callable[..., None]] = attr.ib(
        init=False, factory=list, kw_only=True
    )

    def __attrs_post_init__(self) -> None:
        self._session.register_request(self)

    def __init_extension(
        self, extension_class: Type["AbstractSyrupyExtension"]
    ) -> "AbstractSyrupyExtension":
        return extension_class(test_location=self._test_location)

    @property
    def extension(self) -> "AbstractSyrupyExtension":
        if not self._extension:
            self._extension = self.__init_extension(self._extension_class)
        return self._extension

    @property
    def num_executions(self) -> int:
        return int(self._executions)

    @property
    def executions(self) -> Dict[int, AssertionResult]:
        return self._execution_results

    def use_extension(
        self, extension_class: Optional[Type["AbstractSyrupyExtension"]] = None
    ) -> "SnapshotAssertion":
        """
        Creates a new snapshot assertion fixture with the same options but using
        specified extension class. This does not preserve assertion index or state.
        """
        return self.__class__(
            update_snapshots=self._update_snapshots,
            test_location=self._test_location,
            extension_class=extension_class or self._extension_class,
            session=self._session,
        )

    def assert_match(self, data: "SerializableData") -> None:
        assert self == data

    def _serialize(self, data: "SerializableData") -> "SerializedData":
        return self.extension.serialize(
            data, exclude=self._exclude, matcher=self._matcher
        )

    def get_assert_diff(self) -> List[str]:
        assertion_result = self._execution_results[self.num_executions - 1]
        if assertion_result.exception:
            lines = [
                line
                for lines in traceback.format_exception(
                    assertion_result.exception.__class__,
                    assertion_result.exception,
                    assertion_result.exception.__traceback__,
                )
                for line in lines.splitlines()
            ]
            # Rotate to place exception with message at first line
            return lines[-1:] + lines[:-1]
        snapshot_data = assertion_result.recalled_data
        serialized_data = assertion_result.asserted_data or ""
        diff: List[str] = []
        if snapshot_data is None:
            diff.append(
                gettext("Snapshot '{}' does not exist!").format(
                    assertion_result.snapshot_name
                )
            )
        if not assertion_result.success:
            diff.extend(self.extension.diff_lines(serialized_data, snapshot_data or ""))
        return diff

    def __with_prop(self, prop_name: str, prop_value: Any) -> None:
        setattr(self, prop_name, prop_value)
        self._post_assert_actions.append(lambda: setattr(self, prop_name, None))

    def __call__(
        self,
        *,
        exclude: Optional["PropertyFilter"] = None,
        extension_class: Optional[Type["AbstractSyrupyExtension"]] = None,
        matcher: Optional["PropertyMatcher"] = None,
    ) -> "SnapshotAssertion":
        """
        Modifies assertion instance options
        """
        if exclude:
            self.__with_prop("_exclude", exclude)
        if extension_class:
            self.__with_prop("_extension", self.__init_extension(extension_class))
        if matcher:
            self.__with_prop("_matcher", matcher)
        return self

    def __dir__(self) -> List[str]:
        return ["name", "num_executions"]

    def __eq__(self, other: "SerializableData") -> bool:
        return self._assert(other)

    def _assert(self, data: "SerializableData") -> bool:
        snapshot_location = self.extension.get_location(index=self.num_executions)
        snapshot_name = self.extension.get_snapshot_name(index=self.num_executions)
        snapshot_data: Optional["SerializedData"] = None
        serialized_data: Optional["SerializedData"] = None
        matches = False
        assertion_success = False
        assertion_exception = None
        try:
            snapshot_data = self._recall_data(index=self.num_executions)
            serialized_data = self._serialize(data)
            matches = snapshot_data is not None and serialized_data == snapshot_data
            assertion_success = matches
            if not matches and self._update_snapshots:
                self.extension.write_snapshot(
                    data=serialized_data, index=self.num_executions
                )
                assertion_success = True
            return assertion_success
        except Exception as e:
            assertion_exception = e
            return False
        finally:
            snapshot_created = snapshot_data is None and assertion_success
            snapshot_updated = matches is False and assertion_success
            self._execution_results[self._executions] = AssertionResult(
                snapshot_location=snapshot_location,
                snapshot_name=snapshot_name,
                recalled_data=snapshot_data,
                asserted_data=serialized_data,
                success=assertion_success,
                created=snapshot_created,
                updated=snapshot_updated,
                exception=assertion_exception,
            )
            self._executions += 1
            self._post_assert()

    def _post_assert(self) -> None:
        """
        Restores assertion instance options
        """
        while self._post_assert_actions:
            self._post_assert_actions.pop()()

    def _recall_data(self, index: int) -> Optional["SerializableData"]:
        try:
            return self.extension.read_snapshot(index=index)
        except SnapshotDoesNotExist:
            return None
