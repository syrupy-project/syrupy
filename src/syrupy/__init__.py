import argparse
import sys
from gettext import gettext
from typing import (
    Any,
    ContextManager,
    List,
    Optional,
)

import pytest

from .assertion import SnapshotAssertion
from .constants import DISABLE_COLOR_ENV_VAR
from .exceptions import FailedToLoadModuleMember
from .extensions import DEFAULT_EXTENSION
from .location import PyTestLocation
from .session import SnapshotSession
from .terminal import (
    received_style,
    reset,
    snapshot_style,
)
from .utils import (
    env_context,
    import_module_member,
)

# Global to have access to the session in `pytest_runtest_logfinish` hook
_syrupy: Optional["SnapshotSession"] = None


def __default_extension_option(value: str) -> Any:
    try:
        return import_module_member(value)
    except FailedToLoadModuleMember as e:
        raise argparse.ArgumentTypeError(e)


def pytest_addoption(parser: Any) -> None:
    """
    Exposes snapshot plugin configuration to pytest.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_addoption
    """
    group = parser.getgroup("syrupy")
    group.addoption(
        "--snapshot-update",
        action="store_true",
        default=False,
        dest="update_snapshots",
        help="Update snapshots",
    )
    group.addoption(
        "--snapshot-warn-unused",
        action="store_true",
        default=False,
        dest="warn_unused_snapshots",
        help="Do not fail on unused snapshots",
    )
    group.addoption(
        "--snapshot-details",
        action="store_true",
        default=False,
        dest="include_snapshot_details",
        help="Include details of unused snapshots in the final report",
    )
    group.addoption(
        "--snapshot-default-extension",
        type=__default_extension_option,
        default=DEFAULT_EXTENSION,
        dest="default_extension",
        help="Specify the default snapshot extension",
    )
    group.addoption(
        "--snapshot-no-colors",
        action="store_true",
        default=not sys.stdout.isatty(),
        dest="no_colors",
        help="Disable test results output highlighting",
    )


def __terminal_color(config: Any) -> "ContextManager[None]":
    env = {}
    if config.option.no_colors:
        env[DISABLE_COLOR_ENV_VAR] = "true"

    return env_context(**env)


def pytest_assertrepr_compare(
    config: Any, op: str, left: Any, right: Any
) -> Optional[List[str]]:
    """
    Return explanation for comparisons in failing assert expressions.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_assertrepr_compare
    """
    with __terminal_color(config):
        received_name = received_style("[+ received]")

        def snapshot_name(name: str) -> str:
            return snapshot_style(f"[- {name}]")

        if isinstance(left, SnapshotAssertion):
            assert_msg = reset(f"{snapshot_name(left.name)} {op} {received_name}")
            return [assert_msg] + left.get_assert_diff()
        elif isinstance(right, SnapshotAssertion):
            assert_msg = reset(f"{received_name} {op} {snapshot_name(right.name)}")
            return [assert_msg] + right.get_assert_diff()
    return None


def pytest_sessionstart(session: Any) -> None:
    """
    Initialize snapshot session before tests are collected and ran.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_sessionstart
    """
    session.config._syrupy = SnapshotSession(pytest_session=session)
    global _syrupy
    _syrupy = session.config._syrupy
    session.config._syrupy.start()


def pytest_collection_modifyitems(
    session: Any, config: Any, items: List["pytest.Item"]
) -> None:
    """
    After tests are collected and before any modification is performed.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_collection_modifyitems
    """
    config._syrupy.collect_items(items)


def pytest_collection_finish(session: Any) -> None:
    """
    After collection has been performed and modified.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_collection_finish
    """
    session.config._syrupy.select_items(session.items)


def pytest_runtest_logfinish(nodeid: str) -> None:
    """
    At the end of running the runtest protocol for a single item.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_runtest_logfinish
    """
    global _syrupy
    if _syrupy:
        _syrupy.ran_item(nodeid)


def pytest_sessionfinish(session: Any, exitstatus: int) -> None:
    """
    Finish session run and set exit status.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_sessionfinish
    """
    session.exitstatus |= exitstatus | session.config._syrupy.finish()


def pytest_terminal_summary(
    terminalreporter: Any, exitstatus: int, config: Any
) -> None:
    """
    Add syrupy report to pytest.
    https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_terminal_summary
    """
    with __terminal_color(config):
        terminalreporter.write_sep("-", gettext("snapshot report summary"))
        for line in terminalreporter.config._syrupy.report.lines:
            terminalreporter.write_line(line)


@pytest.fixture
def snapshot(request: Any) -> "SnapshotAssertion":
    return SnapshotAssertion(
        update_snapshots=request.config.option.update_snapshots,
        extension_class=request.config.option.default_extension,
        test_location=PyTestLocation(request.node),
        session=request.session.config._syrupy,
    )
