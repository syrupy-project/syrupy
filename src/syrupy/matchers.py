from typing import (
    TYPE_CHECKING,
    Dict,
    Optional,
    Tuple,
)

from syrupy.extensions.amber import DataSerializer


if TYPE_CHECKING:
    from syrupy.types import (
        PropertyMatcher,
        PropertyPath,
        PropertyValueType,
        SerializableData,
    )


class Repr:
    """
    Easily generate custom representation objects
    """

    def __init__(self, repr_str: str):
        self._repr = repr_str

    def __repr__(self) -> str:
        return self._repr


def path_type(
    mapping: Optional[Dict[str, Tuple["PropertyValueType", ...]]] = None,
    *,
    types: Tuple["PropertyValueType", ...] = (),
    strict: bool = True,
) -> "PropertyMatcher":
    """
    Factory to create a matcher using path and type mapping
    """
    if not mapping and not types:
        raise ValueError("Both mapping and types argument cannot be empty")

    def path_type_matcher(
        data: "SerializableData", path: "PropertyPath"
    ) -> Optional["SerializableData"]:
        path_str = ".".join(str(p) for p, _ in path)
        if mapping:
            for path_to_match in mapping:
                if path_to_match == path_str:
                    for type_to_match in mapping[path_to_match]:
                        if isinstance(data, type_to_match):
                            return Repr(DataSerializer.object_type(data))
                    if strict:
                        raise ValueError(
                            f"{data} at '{path_str}' of type {data.__class__} "
                            f"does not match any of the expected types {types}"
                        )
        for type_to_match in types:
            if isinstance(data, type_to_match):
                return Repr(DataSerializer.object_type(data))
        return data

    return path_type_matcher
