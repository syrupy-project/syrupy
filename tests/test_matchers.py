import datetime
import uuid

import pytest

from syrupy.matchers import (
    PathTypeError,
    path_type,
)


def test_matcher_path_type_noop(snapshot):
    with pytest.raises(PathTypeError, match="argument cannot be empty"):
        path_type()


def test_matches_expected_type(snapshot):
    my_matcher = path_type(
        {"date_created": (datetime.datetime,), "nested.id": (int,)}, types=(uuid.UUID,)
    )
    actual = {
        "date_created": datetime.datetime.now(),
        "nested": {"id": 4},
        "some_uuid": uuid.uuid4(),
    }
    assert actual == snapshot(matcher=my_matcher)


def test_raises_unexpected_type(snapshot):
    kwargs = {
        "mapping": {
            "date_created": (datetime.datetime,),
            "date_updated": (datetime.datetime,),
            "nested.id": (str,),
        },
        "types": (uuid.UUID, int),
    }
    actual = {
        "date_created": datetime.datetime.now(),
        "date_updated": datetime.date(2020, 6, 1),
        "nested": {"id": 4},
        "some_uuid": uuid.uuid4(),
    }
    assert actual == snapshot(matcher=path_type(**kwargs, strict=False))
    with pytest.raises(PathTypeError, match="does not match any of the expected"):
        assert actual == snapshot(matcher=path_type(**kwargs))
