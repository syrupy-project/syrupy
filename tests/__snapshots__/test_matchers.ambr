# name: test_matches_expected_type
  <class 'dict'> {
    'date_created': <class 'datetime'>,
    'nested': <class 'dict'> {
      'id': <class 'int'>,
    },
    'some_uuid': <class 'UUID'>,
  }
---
# name: test_raises_unexpected_type
  <class 'dict'> {
    'date_created': <class 'datetime'>,
    'date_updated': datetime.date(2020, 6, 1),
    'nested': <class 'dict'> {
      'id': <class 'int'>,
    },
    'some_uuid': <class 'UUID'>,
  }
---
